### ISSUE TITLE
Validation errors are shown immediately when dependent value is changed

### ISSUE THREAD

**Issue by abw1023** [jwaliszko/ExpressiveAnnotations#173]
When I have a RequiredIf or AssertThat expression for field B that is dependent on field A, changing field A will immediately show the validation error on B, even though I have not yet attempted to submit the page.  This violates the spirit of jquery.validate, which normally only shows validation errors once a submit is attempted on the page.  I am requesting that expressive annotations be changed to follow the normal behavior of jquery.validate.

I have added the following work-around in validationHelper.validateReferences.  There may be a better way to do this, but this works for my uses.

```
validator = $(form).validate(); // get validator attached to the form
// only validate if validator has been triggered
if (Object.keys(validator.submitted).length == 0) {
    logger.info(typeHelper.string.format('Validator has not been triggered, not validating dependencies of {0}.', name));
    return;
}
```

**Comment by jwaliszko** [comment]
This feature is here by design. If you'd like to mute dependent fields validation, switch it off:
```
<script>
    ea.settings.dependencyTriggers = '';
```
Please refer to the [respective section](https://github.com/jwaliszko/ExpressiveAnnotations#how-to-control-frequency-of-dependent-fields-validation) of the documentation.

**Comment by abw1023** [comment]
`ea.settings.dependencyTriggers = '';`

is not the same thing at all.  This causes dependency validations to be turned off completely, except when submitted.  What I am asking for is the default behavior of jquery.validate, which does not validate until the first submit is attempted, but then updates validations dynamically when valus on the page change.

**Comment by jwaliszko** [comment]
OK, I think I understand the issue now - reopened.

Let me rephrase it by looking at the sample (actual version deployed to http://expressiveannotations.net/). The standard `Required` attribute is applied to the `Age` field. 
![image](https://user-images.githubusercontent.com/1596462/30715003-2a352ccc-9f15-11e7-97e9-56b2ef73366a.png)

The problem you're referring to is, when e.g. some value is selected from the drop-down, and just after that empty value is selected again, while the dependent field `ReasonsForTravel` have been already validated twice (immediately after each selection), the `Age` field is still not validated (not highlighted as required, despite the fact it is empty).

On the other hand, if try to submit the form at the beginning (prior to drop-down manipulations), you'll get the `Age` field validated (highlight as required). 

If instead of using standard `Required` attribute, you change it to EA `RequiredIf("true")`, you'll get the immediate requirement highlight of the `Age` field, even if the form has not been submitted yet.

I prefer not to change this behaviour which is here since a long time (at least not by default), not to surprise other users. Nevertheless we can introduce new configuration setting, e.g `deferValidation`. When `true`, form submission needs to be triggered first, for the field validation to be enabled.

I can do it unless you'd like to contribute with pull request, which is always helpful. Thanks.

**Comment by jwaliszko** [comment]
One more note (just a noisy thinking). I've been looking at the [jquery.validate documentation](https://jqueryvalidation.org/documentation/) and noticed this sentence:

> Before a field is marked as invalid, the validation is lazy: Before submitting the form for the first time, the user can tab through fields without getting annoying messages – they won't get bugged before having the chance to actually enter a correct value

This lazy validation is most likely something you have in mind when it comes to the dependencies validation. But the code `Object.keys(validator.submitted).length == 0`, which works for you, isn't probably quite right, is it? As far as I've debugged through it - the `submitted` object contains information of every invalid field, and is filled up systematically, when a particular field is validated, e.g. on unfocus prior to the form submission. So the form may be not submitted yet, while the `submitted` object contains some fields. It's unclear to me if this condition above is really what you wanted, but I may be wrong - I'm not that familiar with jquery.validate.

**Comment by abw1023** [comment]
I've been digging through the code in jquery.validate and I agree with your statements. `submitted` is not the perfect solution, but I cannot find anything better.  `submitted[element.name]` gets set any time an error is shown and it is never cleared except in a call to `resetForm()`.  It is also used internally on the `focusout`, `keyup` and `click` events to flag whether a validation should be attempted.

So if there are any errors on the submit, then `submitted` will have keys and if there aren't any errors then the form will be submitted so it doesn't matter.  If a field is programmatically validated and is invalid then `submitted` will also get a key, which is what makes it a less than perfect solution.

**Comment by jwaliszko** [comment]
After some consideration I do agree with you, that lazy validation is worth to be turned on by default. All in all user should be able to e.g. tab through the form fields, or should have a chance to provide correct field value, without triggering the immediate validation of dependent fields, and the same seeing the redundant error messages. 

I've created separate [lazy-valid branch](https://github.com/jwaliszko/ExpressiveAnnotations/tree/lazy-valid) to implement this feature. The initial proposal is pushed here: https://github.com/jwaliszko/ExpressiveAnnotations/commit/b0b86af92ef120b4df0fba15992232d89d10447b. You're more than welcome to review that in your spare time, and modify if noticed any issues. 

When doing this changes I've been looking at the latest sources of [jquery.validate](https://cdnjs.cloudflare.com/ajax/libs/jquery-validate/1.17.0/jquery.validate.js) (the script used in the EA web sample is outdated, but I still keep the old version there to be backward compatible).
